### PR TITLE
Static quantization for LSTM (back)

### PR DESCRIPTION
**Pull request by z-a-f**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #32335 Benchmarks for quantized LSTM (static)
* #32074 Static quantization for LSTM (front)
* **#31934 Static quantization for LSTM (back)**

Differential Revision: [D19339398](https://our.internmc.facebook.com/intern/diff/D19339398)

### CODE REVIEW

File: aten/src/ATen/native/RNN.cpp [hunk 1]
```diff
@@ -1168,7 +1280,12 @@ std::tuple<Tensor, Tensor, Tensor> quantized_lstm(
       "dtype is not supported");
 
   std::tuple<Tensor, Tensor, Tensor> results;
-  if (result_dtype == at::kChar || result_dtype == at::kQInt8) {
+  if (result_dtype == at::kQInt8 && !use_dynamic) {
```
Comment (z-a-f) [inline aten/src/ATen/native/RNN.cpp]: @jamesr66a  This is a check for the future -- `if qint8 and not use_dynamic: use static`

**Comment by z-a-f** [comment]
Can I get initial check on this?

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 8461f214:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=8461f214c6da50a1bb4ad336155551faf6ed13de).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [See CircleCI build](https://circleci.com/gh/pytorch/pytorch/4165967) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=81686960) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=81686960))
<details>
<summary>
<code>Jan 08 02:12:15 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o</code>
</summary>

```
Jan 08 02:12:15 ../c10/core/MemoryFormat.h:32:75: warning: ‘c10::MemoryFormat c10::get_contiguous_memory_format()’ is deprecated [-Wdeprecated-declarations] 
Jan 08 02:12:15  #define LEGACY_CONTIGUOUS_MEMORY_FORMAT c10::get_contiguous_memory_format() 
Jan 08 02:12:15                                                                            ^ 
Jan 08 02:12:15 ../aten/src/ATen/native/QuantizedLinear.cpp:251:73: note: in expansion of macro ‘LEGACY_CONTIGUOUS_MEMORY_FORMAT’ 
Jan 08 02:12:15        {weight_contig.size(0)}, weight_contig.options().dtype(at::kInt), LEGACY_CONTIGUOUS_MEMORY_FORMAT); 
Jan 08 02:12:15                                                                          ^ 
Jan 08 02:12:15 ../c10/core/MemoryFormat.h:34:36: note: declared here 
Jan 08 02:12:15  C10_DEPRECATED inline MemoryFormat get_contiguous_memory_format() { 
Jan 08 02:12:15                                     ^ 
Jan 08 02:12:15 [1714/3014] Building CXX object caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o 
Jan 08 02:12:15 FAILED: caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o  
X2 -DTH_HAVE_THREAD -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-write-strings -Wno-unknown-pragmas -Wno-missing-braces -Wno-maybe-uninitialized -fvisibility=hidden -O2 -DCAFFE2_BUILD_MAIN_LIB -DASMJIT_STATIC -std=gnu++14 -MD -MT caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -MF caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o.d -o caffe2/CMakeFiles/torch_cpu.dir/__/aten/src/ATen/native/RNN.cpp.o -c ../aten/src/ATen/native/RNN.cpp 
Jan 08 02:12:15 ../aten/src/ATen/native/RNN.cpp:245:1: error: expected ‘;’ after struct definition 
Jan 08 02:12:15  } 
Jan 08 02:12:15  ^ 
Jan 08 02:12:15 In file included from ../c10/core/Device.h:5:0, 
Jan 08 02:12:15                  from ../c10/core/Allocator.h:6, 
Jan 08 02:12:15                  from ../aten/src/ATen/ATen.h:3, 
Jan 08 02:12:15                  from ../aten/src/ATen/native/RNN.h:3, 
Jan 08 02:12:15                  from ../aten/src/ATen/native/RNN.cpp:1: 
Jan 08 02:12:15 ../aten/src/ATen/native/RNN.cpp: In function ‘std::vector<at::native::{anonymous}::QuantizedCellParamsStatic> at::native::{anonymous}::gather_quantized_params_static(at::TensorList)’: 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by jamesr66a — COMMENTED** [review]
Looks good on the whole, just small comments

File: aten/src/ATen/native/RNN.cpp [hunk 2]
```diff
@@ -195,6 +195,56 @@ struct QuantizedCellParamsDynamic {
   }
 };
 
+struct QuantizedCellParamsStatic {
+  QuantizedCellParamsStatic(const Tensor& _packed_w_ih,
+                            const Tensor& _packed_w_hh,
+                            const Scalar& _output_scale_ih,
+                            const Scalar& _output_scale_hh,
+                            const Scalar& _output_zp_ih,
```
Comment (raghuramank100) [inline aten/src/ATen/native/RNN.cpp]: Do we need an output scale and zero-point? Isnt this fixed?

File: aten/src/ATen/native/RNN.cpp [hunk 3]
```diff
@@ -195,6 +195,52 @@ struct QuantizedCellParamsDynamic {
...
+  Tensor linear_ih(Tensor input) const {
+    const auto kFuncName = "quantized::linear";
+    const auto kOvrldName = "";
+    const std::vector<c10::IValue> output_ih_list =
+        callOp(kFuncName, kOvrldName, input, packed_w_ih,
+               kQ8BitScaleWithRange16, kQ8BitZeroPoint);
+    TORCH_INTERNAL_ASSERT(
+        output_ih_list.size() == 1,
+        "The output vector should have exactly one element");
+    const Tensor output_ih = output_ih_list[0].toTensor();
```
Comment (jamesr66a) [inline aten/src/ATen/native/RNN.cpp]: Not 100% sure but I think this `const` might be preventing an RVO

File: aten/src/ATen/native/RNN.cpp [hunk 4]
```diff
@@ -195,6 +195,52 @@ struct QuantizedCellParamsDynamic {
...
+  Tensor linear_hh(Tensor h) const {
+    const auto kFuncName = "quantized::linear";
+    const auto kOvrldName = "";
+    const std::vector<c10::IValue> output_hh_list =
+        callOp(kFuncName, kOvrldName, h, packed_w_hh,
+               kQ8BitScaleWithRange16, kQ8BitZeroPoint);
+    TORCH_INTERNAL_ASSERT(
+        output_hh_list.size() == 1,
+        "The output vector should have exactly one element");
+    const Tensor output_hh = output_hh_list[0].toTensor();
```
Comment (jamesr66a) [inline aten/src/ATen/native/RNN.cpp]: Same here about `const`

File: aten/src/ATen/native/RNN.cpp [hunk 5]
```diff
@@ -420,7 +480,95 @@ struct LSTMCell : Cell<std::tuple<Tensor, Tensor>, cell_params> {
...
+  static constexpr float kQ32BitScaleWithRange256 = 256.0f / uint32_t(-1);
+  static constexpr int32_t kQZeroPoint = 0;
+
+  hidden_type operator()(
+      const Tensor& input,
+      const hidden_type& hidden,
+      const QuantizedCellParamsStatic& params,
+      bool pre_compute_input = false) const override {
+    const auto& hx = std::get<0>(hidden);
+    const auto& cx = std::get<1>(hidden);
```
Comment (jamesr66a) [inline aten/src/ATen/native/RNN.cpp]: We're pretty deep in type indirection here (auto -> tuple contained type -> using declaration`. Can we just say `const Tensor&` here?

File: aten/src/ATen/native/RNN.cpp [hunk 6]
```diff
@@ -420,7 +480,95 @@ struct LSTMCell : Cell<std::tuple<Tensor, Tensor>, cell_params> {
...
+      auto dqa = qa.dequantize();
+      auto qqa = at::quantize_per_tensor(dqa, out_scale.toDouble(), out_zero_point.toLong(), qtype);
+      return qqa;
+    }
+    if (qa.scalar_type() != at::kQInt32) {
+      return qa;
+    }
+    auto dqa = qa.dequantize();
+    auto qqa = at::quantize_per_tensor(dqa, out_scale.toDouble(), out_zero_point.toLong(), qtype);
+    return qqa;
```
Comment (jamesr66a) [inline aten/src/ATen/native/RNN.cpp]: Can we organize this control flow as
```
if (qa.scalar_type() == at::kQUInt8) {
  ...
} else if (qa.scalar_type() == at::kQInt32) {
  ...
} else {
  ...
}
```
or even as a `switch`.

It took me a second to reason about what's going on here

File: aten/src/ATen/native/RNN.cpp (hunk 3 above)
Comment (jamesr66a) [reply aten/src/ATen/native/RNN.cpp]: Yeah pretty sure that's the case: https://www.bfilipek.com/2017/01/const-move-and-rvo.html

**Pull request closed without merging**